es_source.go

```diff
@@ -87,6 +87,25 @@ func NewElasticsearchSource(config *ElasticsearchConfig) (Source, error) {
 	}, nil
 }
 
+// Create a new ElasticsearchSource from the input object, but add extra SourceFields
+// to include to the new ElasticsearchSource instance.
+func (e *ElasticsearchSource) addExtraFields(extraFields map[string]string) *ElasticsearchSource {
+	sourceFields := make(map[string]string)
+	for k, v := range e.SourceFields {
+		sourceFields[k] = v;
+	}
+	for k, v := range extraFields {
+		sourceFields[k] = v;
+	}
+	return &ElasticsearchSource{
+		ES:            e.ES,
+		Index:         e.Index,
+		GeometryField: e.GeometryField,
+		SourceFields:  sourceFields,
+	}
+}
+
+
```

Comment:
I don't think I've really seen this general pattern before, but it does look to simplify some of the code compared to having to wire everything through a bunch of downstream methods. That said, could probably use slightly different naming (e.g. `withExtraFields` or something similar) so as not to imply that this method mutates the `ElasticsearchSource` object itself.

A more common Go pattern for something similar to this behavior would be to use [struct/interface embedding](https://golang.org/doc/effective_go.html#embedding) and create an actual wrapper struct.

Reply:
I renamed `addExtraFields` to `withExtraFields` and read through the section you linked above.

At the end of the day, I decided to stick with the current code because whatever structure `withExtraFields` returns looks exactly like the structure it takes, interface-wise, so 

I've thought about assuming that we ave an `ElasticsearchSourceTemplate` rather than an `ElasticsearchSource` and then create an actual source for each request, from that template, plus any additional source fields that are requested. -- However, that didn't seem to make things nicer.

So as long as you think there is no risk that the Go Police will show up at my door with an arrest warrant, I'd go with the pattern that we currently have...